txn_writer.go

```diff
@@ -21,121 +25,233 @@ import (
-	Manifest      *Manifest // the manifest to write, if any
-	Lock          *Lock     // the old lock, if any
-	NewLock       gps.Lock  // the new lock, if any
-	SourceManager gps.SourceManager
+	Payload *SafeWriterPayload
 }
 
-// WriteAllSafe writes out some combination of config yaml, lock, and a vendor
-// tree, to a temp dir, then moves them into place if and only if all the write
-// operations succeeded. It also does its best to roll back if any moves fail.
-//
-// This mostly guarantees that dep cannot exit with a partial write that would
-// leave an undefined state on disk.
+// SafeWriterPayload represents the actions SafeWriter will execute when SafeWriter.Write is called.
+type SafeWriterPayload struct {
+	Manifest         *Manifest
+	Lock             *Lock
+	LockDiff         *LockDiff
+	ForceWriteVendor bool
+}
+
+func (payload *SafeWriterPayload) HasLock() bool {
+	return payload.Lock != nil
+}
+
+func (payload *SafeWriterPayload) HasManifest() bool {
+	return payload.Manifest != nil
+}
+
+func (payload *SafeWriterPayload) HasVendor() bool {
+	// TODO(carolynvs) this can be calculated based on if we are writing the lock
+	// init -> switch to newlock
+	// ensure checks existence, why not move that into the prep?
+	return payload.ForceWriteVendor
+}
+
+// LockDiff is the set of differences between an existing lock file and an updated lock file.
+// Fields are only populated when there is a difference, otherwise they are empty.
+// TODO(carolynvs) this should be moved to gps
+type LockDiff struct {
+	HashDiff *StringDiff
+	Add      []LockedProjectDiff
+	Remove   []gps.ProjectRoot
+	Modify   []LockedProjectDiff
+}
+
+func (diff *LockDiff) Format() (string, error) {
+	if diff == nil {
+		return "", nil
+	}
+
+	var buf bytes.Buffer
+
+	if len(diff.Add) > 0 {
+		buf.WriteString("Add: ")
+
+		enc := json.NewEncoder(&buf)
+		enc.SetIndent("", "    ")
+		enc.SetEscapeHTML(false)
+		err := enc.Encode(diff.Add)
+		if err != nil {
+			return "", errors.Wrap(err, "Unable to format LockDiff.Add")
+		}
+	}
+
+	if len(diff.Remove) > 0 {
+		buf.WriteString("Remove: ")
+
+		enc := json.NewEncoder(&buf)
+		enc.SetIndent("", "    ")
+		enc.SetEscapeHTML(false)
+		err := enc.Encode(diff.Remove)
+		if err != nil {
+			return "", errors.Wrap(err, "Unable to format LockDiff.Remove")
+		}
+	}
+
+	if len(diff.Modify) > 0 {
+		buf.WriteString("Modify: ")
+
+		enc := json.NewEncoder(&buf)
+		enc.SetIndent("", "    ")
+		enc.SetEscapeHTML(false)
+		err := enc.Encode(diff.Modify)
+		if err != nil {
+			return "", errors.Wrap(err, "Unable to format LockDiff.Modify")
+		}
+	}
+
+	return buf.String(), nil
+}
+
+// LockedProjectDiff contains the before and after snapshot of a project reference.
+// Fields are only populated when there is a difference, otherwise they are empty.
+// TODO(carolynvs) this should be moved to gps
+type LockedProjectDiff struct {
+	Name       gps.ProjectRoot `json:"name"`
+	Repository *StringDiff     `json:"repo,omitempty"`
```

Comment:
@sdboyer Do you want the json name changed as well? Right now it's `repo` which matches what is output in the lock file.